### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by ashwath462**

### CODE REVIEW

**Review by piyushmalviya13 — CHANGES_REQUESTED** [review]
Look into the comments.

File: utility/constants.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { question } from "readline-sync";
+export function input(message:string){
+    const userInput = question(message);
+    return userInput;
+}
+
+export const onlyLettersAndSpaces = /^[A-Za-z\s]*$/;
```
[the remark this answers is not in the document]
Comment (ashwath462) [reply utility/constants.ts:7]: Done

File: model/File.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        const rollNumber = userInput.getRollNumber();
+        let allUsers = this.getAllUsers();
+        const indexToRemove = allUsers.findIndex((user:any) => user.rollNumber === rollNumber);
+        if (indexToRemove !== -1) {
+            throw new Error(`User with roll number ${rollNumber} already exist`);
+        }
+        allUsers.push(userInput);
+        let hashedUserData:any = '';
+        this.sortUsers(allUsers);
+        allUsers.map((user:any)=>{
```
[the remark this answers is not in the document]
Comment (ashwath462) [reply model/File.ts]: done

File: model/File.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        let allUsers:any = data.split(',').slice(1);
+
+        allUsers = allUsers.map((element:any)=>{
+            let bufferObj = Buffer.from(element, "base64"); 
+            return JSON.parse(bufferObj?.toString("utf8"));
+        })
+        return allUsers;
+    }
+
+    public deleteUser(rollNumber:number){
```
[the remark this answers is not in the document]
Comment (ashwath462) [reply model/File.ts]: Done.

File: model/File.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    private constructor(){};
+
+    public static getInstance():File{
+        if(!File.instance){
+            File.instance = new File();
+        }
+        return File.instance;
+    }
+
+    public sortUsers(allUsers:any){
```
[the remark this answers is not in the document]
Comment (ashwath462) [reply model/File.ts]: Done.

File: assignment2.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    default:
+      console.log("Invalid input. Try Again!\n");
+  }
+}
+console.log(file.getAllUsers());
+
+
+function addUserDetails() {
+  console.log("Kindly enter the following details about user : ");
+  while (true) {
```
[the remark this answers is not in the document]
Comment (ashwath462) [reply assignment2.ts:51]: Corrected.

File: assignment2.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+const file = File.getInstance();
+console.clear();
+console.log(
+  "Menu : \n1. Add User details. \n2. Display User details. \n3. Delete User detils. \n4. Save User details. \n5. Exit"
+);
+while (flag) {
+  const inputChoice = input("Enter a choice form menu (1-5) : ");
+  switch (inputChoice) {
+    case "1":
+      user = addUserDetails();
```
[the remark this answers is not in the document]
Comment (ashwath462) [reply assignment2.ts]: Correction done.

File: assignment2.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      break;
+    case "2":
+      if (user != null) {
+        user.displayUserDetails();
```
[the remark this answers is not in the document]
Comment (ashwath462) [reply assignment2.ts]: Implemented in user class.

File: assignment2.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+      } else {
+        console.log("No user details found");
+      }
+      break;
+    case "3":
+      const rollNumber = validateRollNumber(
+        input("Enter roll number to be deleted : ")
+      );
+      file.deleteUser(rollNumber);
```
[the remark this answers is not in the document]
Comment (ashwath462) [reply assignment2.ts]: Done.

File: assignment2.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  validateRollNumber,
+  validateCourses,
+} from "./utility/utils";
+import { input } from "./utility/constants";
+import { File } from "./model/File";
+
+
+
+let flag = true;
+let user: User | null = null;
```
[the remark this answers is not in the document]
Comment (ashwath462) [reply assignment2.ts]: Implemented

File: assignment2.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+console.log(
+  "<-------------------------------------- Welcome to our application -------------------------------------->"
+);
+const file = File.getInstance();
```
[the remark this answers is not in the document]
Comment (ashwath462) [reply assignment2.ts]: Done.

File: tests/utility/file.test.ts [hunk 11]
```diff
@@ -0,0 +1,106 @@
+import { expect, test, vi } from "vitest";
```
[the remark this answers is not in the document]
Comment (ashwath462) [reply tests/utility/file.test.ts]: Added screenshot

File: model/File.ts [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+  public static getInstance(): File {
+    if (!File.instance) {
+      File.instance = new File();
+    }
+    return File.instance;
+  }
+
+  public saveData(hashedUserData: string) {
+    if(hashedUserData.length>=1){
+        fsPromise.writeFileSync("./database/db.txt", hashedUserData);
```
Comment (piyushmalviya13) [inline model/File.ts:18]: Make the file name a constant and use that.

File: assignment2.ts [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        input("Enter roll number to be deleted : "),
+        true
+      );
+      userStore.deleteUser(rollNumber);
+      break;
+    case "4":
+      userStore.saveData();
+      break;
+    case "5":
+      console.log("Have a good day!");
```
Comment (piyushmalviya13) [inline assignment2.ts:54]: Problem statement says to ask user if they want to save the changes done to data.

File: assignment2.ts [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+  switch (inputChoice) {
+    case "1":
+      addUserDetails();
+      break;
+    case "2":
+      console.clear();
+      const inputChoice = sortUserChoice(input("Sort user by : \n1. Name \n2. Roll Number \n3. Age \n4. Address\n"));
+      let sortOrder:any = input('1. Ascending\n2. Descending\n');
+      console.clear();
+      if(sortOrder == '1' || sortOrder == '2'){
```
Comment (piyushmalviya13) [inline assignment2.ts:33]: What is this logic doing?

File: utility/utils.ts [hunk 15]
```diff
@@ -1,3 +1,98 @@
...
+    throw new Error("Do not enter a decimal value or 0");
+  }
+  if (rollNumber <= 0 || rollNumber >= 100000) {
+    throw new Error("Age can only be between 1-99999");
+  }
+  if (toDelete) {
+    return rollNumber;
+  }
+  const data = new UsersData().getUserData();
+  const indexToRemove = data.findIndex(
```
Comment (piyushmalviya13) [inline utility/utils.ts:53]: Why is the variable named indexToRemove?

File: utility/utils.ts [hunk 16]
```diff
@@ -1,3 +1,98 @@
...
+  }
+  if (age <= 0 || age >= 100) {
+    throw new Error("Age can only be between 1-99");
+  }
+  return age;
+}
+
+export function validateRollNumber(
+  userInput: any,
+  toDelete: boolean = false
```
Comment (piyushmalviya13) [inline utility/utils.ts:37]: It's better to move logic to find is roll number exists in another function. 
Validate function shouldn't hold this logic.

File: model/File.ts [hunk 17]
```diff
@@ -0,0 +1,24 @@
+import * as fsPromise from "fs";
+import { User } from "./User";
+
+export class File {
+  private static instance: File;
+
+  private constructor() {}
```
Comment (piyushmalviya13) [inline model/File.ts:7]: Is this required?

File: model/UsersData.ts [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+    let allUsers: any = data.split(",").slice(1);
+
+    allUsers = allUsers.map((element: any) => {
+      let bufferObj = Buffer.from(element, "base64");
+      return JSON.parse(bufferObj?.toString("utf8"));
+    });
+    this.usersData = allUsers;
+  }
+
+  public getUserData() {
```
Comment (piyushmalviya13) [inline model/UsersData.ts:20]: Need of getter for a public member userData?

File: model/UsersData.ts [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+    return this.usersData;
+  }
+
+  public displayUserDetails(inputChoice:number,sortOrder:boolean) {
+    if(inputChoice == 1){
+      this.usersData = this.sortByName(this.usersData,sortOrder);
```
Comment (piyushmalviya13) [inline model/UsersData.ts:26]: While displaying data the sorting is done on the original array of users, which will reflect on storing the file. 
It should only be displayed in sorted manner, the original order (by roll number) should be retained.

File: model/UsersData.ts [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+    // Display column headers
+    console.log(
+      columnWidths
+        .map((width, index) => headers[index].padEnd(width))
+        .join(" | ")
+    );
+    // Display separator line
+    console.log(
+      "-".repeat(columnWidths.reduce((sum, width) => sum + width + 3, 0))
+    );
```
Comment (piyushmalviya13) [inline model/UsersData.ts:58]: Move this logic to utility file. It has nothing to do with this class.

File: model/UsersData.ts [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+  public sortByRoleNumber(usersData: any, descOrder: boolean = false) {
+    usersData.sort((a: any, b: any) => {
+      if (descOrder) return b.rollNumber - a.rollNumber;
+      return a.rollNumber - b.rollNumber;
+    });
+    return usersData;
+  }
+
+  public sortByAge(usersData: any, descOrder: boolean = false) {
+    usersData.sort((a: any, b: any) => {
```
Comment (piyushmalviya13) [inline model/UsersData.ts:101]: Refactor this code.

**Review by piyushmalviya13 — CHANGES_REQUESTED** [review]
Look into the comments

File: model/File.ts (hunk 12 above)
Comment (ashwath462) [reply model/File.ts:18]: Done.

File: assignment2.ts (hunk 14 above)
Comment (ashwath462) [reply assignment2.ts:33]: Discussed.

File: model/File.ts (hunk 17 above)
Comment (ashwath462) [reply model/File.ts:7]: To prevent object formation constructor must be private.

File: model/UsersData.ts (hunk 19 above)
Comment (ashwath462) [reply model/UsersData.ts:26]: Done

File: model/UsersData.ts (hunk 20 above)
Comment (ashwath462) [reply model/UsersData.ts:58]: Done

File: model/UsersData.ts (hunk 18 above)
Comment (ashwath462) [reply model/UsersData.ts:20]: Resolved.

File: utility/utils.ts (hunk 16 above)
Comment (ashwath462) [reply utility/utils.ts:46]: Done

File: model/UsersData.ts (hunk 21 above)
Comment (ashwath462) [reply model/UsersData.ts]: Done

File: utility/utils.ts (hunk 15 above)
Comment (ashwath462) [reply utility/utils.ts]: Changed.

File: assignment2.ts (hunk 13 above)
Comment (ashwath462) [reply assignment2.ts:67]: Implemented.

**Review by piyushmalviya13 — APPROVED** [review]
LGTM

**Pull request merged**